assets/dns/daemonset.yaml

```diff
@@ -44,14 +44,56 @@ spec:
           timeoutSeconds: 5
           successThreshold: 1
           failureThreshold: 5
+      - name: dns-node-resolver
+        # image is set at runtime
+        imagePullPolicy: IfNotPresent
+        securityContext:
+          privileged: true
+        volumeMounts:
+        - name: hosts-file
+          mountPath: /etc/hosts
+        # env NAMESERVER and CLUSTER_DOMAIN are set at runtime
+        command:
+        - /bin/bash
+        - -c
+        - |
+          #!/bin/bash
+          set -euo pipefail
+
+          REGISTRY_SERVICE="docker-registry.default.svc"
+          OPENSHIFT_MARKER="openshift-generated-node-resolver"
+          HOSTS_FILE="/etc/hosts"
+          TEMP_FILE="/etc/hosts.tmp"
+
+          while true; do
+            # Fetch registry service IP from cluster dns if present
+            ips=($(dig @"${NAMESERVER}" +short "${REGISTRY_SERVICE}.${CLUSTER_DOMAIN}"))
+            if [[ ("$?" -eq 0) && ("${#ips[@]}" -ne 0) ]]; then
+              # Build a new hosts file from /etc/hosts with our custom entries filtered out
+              grep -v "${OPENSHIFT_MARKER}" "${HOSTS_FILE}" > "${TEMP_FILE}"
+
+              # Append resolver entries for registry
+              for ip in "${ips[@]}"; do
+                echo "${ip} ${REGISTRY_SERVICE} # ${OPENSHIFT_MARKER}" >> "${TEMP_FILE}"
+              done
+
+              # Replace /etc/hosts with our modified version
+              cp -f "${TEMP_FILE}" "${HOSTS_FILE}"
```

Comment:
It's traditional to use `mv` in order to perform the replacement atomically.  For `/etc/hosts`, we also want to take care to preserve attributes including SELinux context, by doing something along the lines of the following:
```bash
            if [[ "$?" -eq 0 && "${#ips[@]}" -ne 0 ]]; then
              # Make a temporary file with the old hosts file's attributes.
              cp -f --attributes-only "${HOSTS_FILE}" "${TEMP_FILE}"
              
              # Write to $TEMP_FILE using > and >> as usual.
              # ...
              
              # Atomically replace /etc/hosts with the temporary file.
              mv -f "${TEMP_FILE}" "${HOSTS_FILE}"
            fi
```
Based on item 9 at http://www.pixelbeat.org/docs/unix_file_replacement.html.

Reply:
Tried using `mv` from temp file to /etc/hosts before but looks like that is not supported. Got this error: 
`mv: cannot move '/etc/hosts.tmp' to '/etc/hosts': Device or resource busy`
Searching online, it was recommended to use `cp -f` for this use case.

I will copy initial /etc/hosts file attributes when creating the new temporary file to preserve selinux context.

Reply:
Oh, probably because it's a bind mount—atomicity might not be possible here.

Reply:
Related issue: https://github.com/moby/moby/issues/17190